src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs

```diff
@@ -0,0 +1,395 @@
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+                            Justification = "This calls WriteEvent with all primitive arguments which is safe. Primitives are always serialized properly.")]
+        public void InstrumentPublished(string sessionId, string meterName, string? meterVersion, string instrumentName, string instrumentType, string? unit, string? description)
+        {
+            WriteEvent(11, sessionId, meterName, meterVersion ?? "", instrumentName, instrumentType, unit ?? "", description ?? "");
+        }
+
+        /// <summary>
+        /// Called when the EventSource gets a command from a EventListener or ETW.
+        /// </summary>
+        [NonEvent]
+        protected override void OnEventCommand(EventCommandEventArgs command)
+        {
+            lock (this)
+            {
+                _handler.OnEventCommand(command);
+            }
+        }
+
+        // EventSource assumes that every method defined on it represents an event.
+        // Methods that are declared explicitly can use the [NonEvent] attribute to opt-out but
+        // lambdas can't. Putting all the command handling logic in this nested class
+        // is a simpler way to opt everything out in bulk.
+        private class CommandHandler
+        {
+            private AggregationManager? _aggregationManager;
+            private string _sessionId = "";
+
+            public void OnEventCommand(EventCommandEventArgs command)
+            {
+                try
+                {
+#if OS_SUPPORT_ATTRTIBUTES
+                    if (OperatingSystem.IsBrowser())
+                    {
+                        // AggregationManager uses a dedicated thread to avoid losing data for apps experiencing threadpool starvation
+                        // and browser doesn't support Thread.Start()
+                        //
+                        // This limitation shouldn't really matter because browser also doesn't support out-of-proc EventSource communication
+                        // which is the intended scenario for this EventSource. If it matters in the future AggregationManager can be
+                        // modified to have some other fallback path that works for browser.
+                        Logger.Error("", "System.Diagnostics.Metrics EventSource not supported on browser", "");
+                        return;
+                    }
+#endif
+                    if (command.Command == EventCommand.Update || command.Command == EventCommand.Disable ||
+                        command.Command == EventCommand.Enable)
+                    {
+                        if (_aggregationManager != null)
+                        {
+                            _aggregationManager.Dispose();
+                            _aggregationManager = null;
+                            Logger.Message($"Previous session with id {_sessionId} is stopped");
```

Comment:
If we're here, are these events guaranteed to be enabled?  If not, we're potentially doing formatting work that'll just be throwaway.

Reply:
It is not guaranteed the events are enabled, but these code paths should run very rarely and I wouldn't suggest writing extra code to optimize it.